tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/EnforceRateLimitingPlan.scala

```diff
@@ -0,0 +1,145 @@
+package com.linagora.tmail.mailets
+
+import com.google.common.base.Preconditions
+import com.google.common.collect.ImmutableList
+import com.linagora.tmail.rate.limiter.api.{OperationLimitations, RateLimitationPlanRepository, RateLimitingPlanId, RateLimitingPlanNotFoundException, RateLimitingPlanUserRepository}
+import org.apache.james.core.{MailAddress, Username}
+import org.apache.james.lifecycle.api.LifecycleUtil
+import org.apache.james.rate.limiter.api.{AcceptableRate, RateExceeded, RateLimiterFactory, RateLimitingResult, Rule, Rules}
+import org.apache.james.transport.mailets.{KeyPrefix, PrecisionParsingUtil}
+import org.apache.mailet.Mail
+import org.apache.mailet.base.GenericMailet
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import java.time.Duration
+import javax.inject.Inject
+import scala.jdk.CollectionConverters._
+import scala.util.Using
+
+class EnforceRateLimitingPlan @Inject()(planRepository: RateLimitationPlanRepository,
+                                        planUserRepository: RateLimitingPlanUserRepository,
+                                        rateLimiterFactory: RateLimiterFactory) extends GenericMailet {
+
+  private var operationLimitation: String = _
+  private var exceededProcessor: String = _
+  private var rateLimitSender: Boolean = _
+  private var rateLimitRecipient: Boolean = _
+  private var planRateLimiterResolver: PlanRateLimiterResolver = _
+
+  override def init(): Unit = {
+    exceededProcessor = getInitParameter("exceededProcessor", Mail.ERROR)
+    rateLimitSender = getInitParameter("rateLimitSender", false)
+    rateLimitRecipient = getInitParameter("rateLimitRecipient", false)
```

Comment:
I think we need to rework this API.

Maybe we coult ask our mailet to "Apply delivery limits" or "Apply relay limits" and the mailet code would know if we need to rate limit the sender or the recipients depending on the case. So that the admin writing the conf don't need to know...